auth.secret_name provided but still fails with "authentication required" on Docker Hub private repo / cosign

### Discussed in https://github.com/sse-secure-systems/connaisseur/discussions/573

<div type='discussions-op-text'>

<sup>Originally posted by **funkypenguin** March  7, 2022</sup>
Hey folks!

I'm trying to use a cosign validator against my private Docker Hub repos, but I've not been able to get connaisseur to authenticate - so I get errors like this when testing with `kubectl run`:

```
Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request: Unexpected Cosign exception for image "docker.io/myord/curl:latest": Error: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:gergorg/curl Type:repository]]
main.go:46: error during command execution: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]
```

My validator is defined like this:

```
  - name: default
    type: cosign  # or other supported validator (e.g. "cosign")
    # host: notary.docker.io # in case of notaryv1, configure the notary server to be used
    auth: 
      secret_name: registry-credentials  # reference a k8s secret in the form required by the validator type (check the docs)
    trust_roots:
    # # the `default` key is used if no key is specified in image policy
    - name: default
      key: |  # enter your public key below (my note : the one below is a dummy one used for dev)
        -----BEGIN PUBLIC KEY-----
        MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEizCL5gnS98PS929yPNo61x7MSuhs
        g4qfWtLxfaRQfy2YPhZ2//y+AeaGLU9Y98jQHyZqLON4F+9DFJqX2RbwFQ==
        -----END PUBLIC KEY-----
    #  # or (only for notaryv1 validator)
    #  username: myuser
    #  password: mypass
```

And my I've confirmed that the credentials are correctly mounted into the pod container at `/app/connaisseur-config/default/.docker/config.json`.

When I set `LOGLEVEL=DEBUG`, I see the following output from connaisseur:

```
[2022-03-07 01:52:23,831] INFO: {'message': 'Unexpected Cosign exception for image "docker.io/myorg/curl:latest": Error: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\nmain.go:46: error during command execution: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\n.', 'context': {'trust_data_type': 'dev.cosignproject.cosign/signature', 'stderr': 'Error: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\nmain.go:46: error during command execution: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\n', 'image': 'docker.io/myorg/curl:latest', 'trust_root': 'default', 'detection_mode': False}}
[2022-03-07 01:52:23,832] DEBUG: No alerting configuration file found.
[2022-03-07 01:52:23,832] ERROR: {'message': 'Unexpected Cosign exception for image "docker.io/myorg/curl:latest": Error: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\nmain.go:46: error during command execution: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\n.', 'context': {'trust_data_type': 'dev.cosignproject.cosign/signature', 'stderr': 'Error: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\nmain.go:46: error during command execution: GET https://index.docker.io/v2/myorg/curl/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:myorg/curl Type:repository]]\n', 'image': 'docker.io/myorg/curl:latest', 'trust_root': 'default', 'detection_mode': False, 'user': 'system:admin', 'operation': 'CREATE', 'kind': 'Pod', 'name': 'test', 'namespace': 'sandbox'}}
```

Any ideas? :)

Thanks,
D
</div>

--- a/connaisseur/validators/cosign/cosign_validator.py
+++ b/connaisseur/validators/cosign/cosign_validator.py
@@ -226,10 +226,12 @@ class CosignValidator(ValidatorInterface):
             *(["--k8s-keychain"] if self.k8s_keychain else []),
             image,
         ]
+        env = self.__get_envs()
+        env.update(env_vars)
 
         with subprocess.Popen(  # nosec
             cmd,
-            env=self.__get_envs().update(env_vars),
+            env=env,
             stdin=subprocess.PIPE,
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE,

--- a/helm/Chart.yaml
+++ b/helm/Chart.yaml
@@ -2,8 +2,8 @@ apiVersion: v2
 name: connaisseur
 description: Helm chart for Connaisseur - a Kubernetes admission controller to integrate container image signature verification and trust pinning into a cluster.
 type: application
-version: 1.3.0
-appVersion: 2.5.0
+version: 1.3.1
+appVersion: 2.5.1
 keywords:
   - container image
   - signature

--- a/helm/values.yaml
+++ b/helm/values.yaml
@@ -1,7 +1,7 @@
 # configure Connaisseur deployment
 deployment:
   replicasCount: 3
-  image: securesystemsengineering/connaisseur:v2.5.0
+  image: securesystemsengineering/connaisseur:v2.5.1
   imagePullPolicy: IfNotPresent
   # imagePullSecrets contains an optional list of Kubernetes Secrets, in Connaisseur namespace,
   # that are needed to access the registry containing Connaisseur image.

